[Bug] Unblock priority spread under Psychic Terrain

acelynnzhang opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-  [Bug]: If the PR is primarily a bug fix
-  [Move]: If a move has new or changed functionality
-  [Ability]: If an ability has new or changed functionality
-  [Item]: For new or modified items
-  [Mystery]: For new or modified Mystery Encounters
-  [Test]: If the PR is primarily adding or modifying tests
-  [UI/UX]: If the PR is changing UI/UX elements
-  [Audio]: If the PR is adding or changing music/sfx
-  [Sprite]: If the PR is adding or changing sprites
-  [Balance]: If the PR is related to game balance
-  [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-  [Dev]: If the PR is primarily changing something pertaining to development (lefthook hooks, linter rules, etc.)
-  [i18n]: If the PR is primarily adding/changing locale keys or key usage (may come with an associated locales PR)
-  [Docs]: If the PR is adding or modifying documentation (such as tsdocs/code comments)
-  [GitHub]: For changes to GitHub workflows/templates/etc
-  [Misc]: If no other category fits the PR
-->

<!--
Make sure that this PR is not overlapping with someone else's work
Please try to keep the PR self-contained (and small!)
-->

## What are the changes the user will see?
<!-- Summarize what are the changes from a user perspective on the application -->
Spread moves will work with abilities that give priority on Psychic Terrain.
## Why am I making these changes?
<!--
Explain why you decided to introduce these changes
Does it come from an issue or another PR? Please link it
Explain why you believe this can enhance user experience
-->
<!--
If there are existing GitHub issues related to the PR that would be fixed,
you can add "Fixes #[issue number]" (ie: "Fixes #1234") to link an issue to your PR
so that it will automatically be closed when the PR is merged.
-->
Fixes #6056

## What are the changes from a developer perspective?
<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->
The function `isMoveTerrainCancelled` in `terrain.ts` has an extra check to see if the move being used is a spread or field move. A file for testing Psychic Terrain has been added.
## Screenshots/Videos
<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->
Before:

https://github.com/user-attachments/assets/c6a056ff-22c8-4365-b5bb-27a4b6853903

After:

https://github.com/user-attachments/assets/ca903325-4f89-4314-8e01-fa32ae22add9


## How to test the changes?
<!--
How can a reviewer test your changes once they check out on your branch?
Did you make use of the `src/overrides.ts` file?
Did you introduce any automated tests?
Do the reviewers need to do something special in order to test your changes?
-->
`const overrides = {ABILITY_OVERRIDE: AbilityId.PRANKSTER, MOVESET_OVERRIDE: [MoveId.RAIN_DANCE, MoveId.DARK_VOID, MoveId.PSYCHIC_TERRAIN], OPP_MOVESET_OVERRIDE: [MoveId.SPLASH], OPP_LEVEL_OVERRIDE: 10, OPP_SPECIES_OVERRIDE: SpeciesId.TORCHIC} satisfies Partial<InstanceType<typeof DefaultOverrides>>;`

Use Psychic Terrain and then Rain Dance.

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [ ] ~~Have I made sure that any UI change works for both UI themes (default and legacy)?~~

Bertie690 commented on src/data/terrain.ts:
```diff
@@ -57,10 +58,22 @@ export class Terrain {
     return 1;
   }
 
+  isSpreadOrFieldMove(move: Move): boolean {
+    switch (move.moveTarget) {
+      case MoveTarget.ALL_ENEMIES:
+      case MoveTarget.ALL_NEAR_ENEMIES:
+      case MoveTarget.ALL_OTHERS:
+      case MoveTarget.ALL_NEAR_OTHERS:
+      case MoveTarget.BOTH_SIDES:
+        return true;
+    }
+    return false;
+  }
```
This should be a method on `Move`, not inside the terrain file.

Also pardon my poor memory, but I recall seeing something _vaguely_ like a "is this a spread move" function in there already. Might be worth checking

Bertie690 replied:
That would be good, yes.
I'll check the end result in a min

Bertie690 commented on src/data/terrain.ts:
```diff
@@ -57,10 +58,22 @@ export class Terrain {
     switch (this.terrainType) {
       case TerrainType.PSYCHIC:
-        if (!move.hasAttr("ProtectAttr")) {
+        if (!move.hasAttr("ProtectAttr") && !this.isSpreadOrFieldMove(move)) {
           // Cancels move if the move has positive priority and targets a Pokemon grounded on the Psychic Terrain
           return (
             move.getPriority(user) > 0 &&
```
Can we combine this with the return thing into 1 big check?
(The protection check can also be removed since all protection moves are single or ally targeted AFAIK)

Bertie690 reviewed (changes requested):
Preliminary stuff - will send a patch with the test changes soon

acelynnzhang commented:
I have moved the spread check to move.utils and made use of the existing isFieldMove. I have also condensed the return on the psychic terrain to 1 big check and removed the protection check.

Bertie690 commented on src/data/moves/move-utils.ts:
```diff
@@ -27,6 +27,17 @@ export function isFieldTargeted(move: Move): boolean {
   return false;
 }
 
+export function isSpreadMove(move: Move): boolean {
```
```suggestion
/**
 * Return whether the move is a spread move.
 *
 * Examples include:
 * - Moves targeting all adjacent pokemon (like Surf)
 * - Moves targeting all adjacent enemies (like Air Cutter)
 */
export function isSpreadMove(move: Move): boolean {
```

Bertie690 commented on src/data/terrain.ts:
```diff
@@ -60,13 +61,13 @@ export class Terrain {
-            user.getOpponents(true).some(o => targets.includes(o.getBattlerIndex()) && o.isGrounded())
-          );
-        }
+        // Cancels move if the move has positive priority and targets a Pokemon grounded on the Psychic Terrain
+        return (
+          !isFieldTargeted(move) &&
+          !isSpreadMove(move) &&
+          move.getPriority(user) > 0 &&
+          user.getOpponents(true).some(o => targets.includes(o.getBattlerIndex()) && o.isGrounded())
+        );
```
```suggestion
        // Cf https://bulbapedia.bulbagarden.net/wiki/Psychic_Terrain_(move)#Generation_VII
        // Psychic terrain will cancel a move if:
        return (
          // ...the move is neither field nor spread-targeted,
          !isFieldTargeted(move) &&
          !isSpreadMove(move) &&
          // ...the move has positive final priority,
          move.getPriority(user) > 0 &&
          // ...and is targeting at least 1 grounded opponent.
          user.getOpponents(true).some(o => targets.includes(o.getBattlerIndex()) && o.isGrounded())
        );
```

SirzBenjie replied:
According to bulbapedia, it seems that perish song _is_ supposed to be blocked by psychic terrain if called with priority? Maybe not. Idk, but either way idt it's necessary to account for it.

acelynnzhang commented:
I have added the comments as suggested. I will try to be better with comments in the future.

SirzBenjie commented:
Also, don't worry about failing tests, let me fix.

SirzBenjie commented:
> Oh my bad, I didn't see this

Np.

Bertie690 commented:
> These are just additions that are being suggested to further improve clarity -- not a deficiency on anything that you wrote.

Benjie was correct - this was never the intended message to convey. 
Keep up the good work! :)
